### PR TITLE
Implement table management

### PR DESCRIPTION
**Pull request by aletor123**

### CODE REVIEW

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 1]
```diff
@@ -50,12 +110,199 @@ export class AnimeTableComponent implements OnInit {
...
+            field: AnimeFilterOptions.Type,
+            filterType: FilterType.In,
+            value: types,
+          });
+        }
+        this.listManager.updateFilters(filterOptions);
+      },
+    );
+
+    this.listManager.pagePagination$.pipe(
```
Comment (gevorg-gevorgyan-saritasa) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:154]: Please try to avoid such amount of subscriptions. You can combine your filer / search / pagination / sorting into one `Observable` inside constructor to detect their changes, and then pipe it to fetch your anime list

**Review by gevorg-gevorgyan-saritasa — CHANGES_REQUESTED** [review]

File: apps/angular/src/app/features/anime/table/anime-table.component.ts (hunk 1 above)
Comment (aletor123) [reply apps/angular/src/app/features/anime/table/anime-table.component.ts:138]: I reduced number of subscriptions. I'm already combining all this parameters in ListManager that can be reused in other places where pagination is needed

**Review by artem-sotnichenko-saritasa — APPROVED** [review]

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 2]
```diff
@@ -13,32 +48,55 @@ import { AnimeService } from '../../../../core/services/anime.service';
...
+    private formBuilder: FormBuilder,
+    private activeRoute: ActivatedRoute,
+    private router: Router,
     animeService: AnimeService,
   ) {
-    this.animeList$ = animeService.getAnimeList().pipe(
-      map(animePaginatedList => animePaginatedList.items),
-      shareReplay({ refCount: true, bufferSize: 1 }),
+    this.initialParams = this.getParamsFromQuery(this.activeRoute.snapshot.queryParamMap);
+    this.listManager = new ListManager(this.initialParams);
```
[the remark this answers is not in the document]
Comment (aletor123) [reply apps/angular/src/app/features/anime/table/anime-table.component.ts:90]: I removed queryParams$ but list manager initialized in constructor because i need to take initial parameters from query

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 3]
```diff
@@ -53,15 +111,196 @@ export class AnimeTableComponent implements OnInit, OnDestroy {
...
+          });
+        }
+        this.queryParams$.next(queryParams);
       },
     ));
   }
 
+  /** @inheritDoc */
+  public ngAfterViewInit(): void {
+    const { sortOptions, filtersOptions } = this.getParamsFromQuery(this.activeRoute.snapshot.queryParamMap);
```
[the remark this answers is not in the document]
Comment (aletor123) [reply apps/angular/src/app/features/anime/table/anime-table.component.ts:171]: I'm doing it in `ngAfterViewInit` because i need to access  `tableSort` that is child component. And it's initialized only in `ngAfterViewInit`

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 4]
```diff
@@ -1,9 +1,46 @@
...
 import { UntilDestroy, untilDestroyed } from '@ngneat/until-destroy';
 
+import { isSortDirection, SortDirection, SortOptions } from '@js-camp/core/models/sortOptions';
+
 import { AnimeService } from '../../../../core/services/anime.service';
+import { ListManager, ListManagerInitParams } from '../../../../core/utils/list-manager';
+import { matSortToSortOptions } from '../../../../core/utils/table';
+import { getNumberQueryParameter } from '../../../../core/utils/queryParameters';
+
+interface AnimeFiltersType {
```
Comment (artem-sotnichenko-saritasa) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:20]: We usually add suffix `FormData` for form type name, i.e. `AnimeFiltersFormData`

**Review by artem-sotnichenko-saritasa — APPROVED** [review]

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 5]
```diff
@@ -50,12 +108,186 @@ export class AnimeTableComponent implements OnInit {
...
+        if (search !== null && search !== undefined) {
+          filterOptions.push({
+            field: AnimeFilterOptions.Search,
+            filterType: FilterType.Exact,
+            value: search,
+          });
+        }
+
+        const types = filterValues[AnimeFilterOptions.Type];
+        if (types !== null && types !== undefined) {
```
Comment (timofei-maximov-saritasa) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:127]: FYI, you can use `types != null` for `types: null | undefined`

**Review by timofei-maximov-saritasa — APPROVED** [review]

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 6]
```diff
@@ -1,9 +1,46 @@
...
+  SEARCH = 'search',
+  TYPE = 'type',
+  PAGE = 'page',
+  PAGE_SIZE = 'pageSize',
+  ORDERING_FIELD = 'orderingField',
+  ORDERING_DIRECTION = 'orderingDirection',
+}
+
+interface QueryParameters {
+  [key: string]: string;
```
Comment (pavel-romanov8) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:40]: You can use `Record<string, string>` for that instead of creating a separate interface

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 7]
```diff
@@ -1,9 +1,46 @@
...
+}
+
+/** Anime table query parameters. */
+enum TableQueryParams {
+  SEARCH = 'search',
+  TYPE = 'type',
+  PAGE = 'page',
+  PAGE_SIZE = 'pageSize',
+  ORDERING_FIELD = 'orderingField',
+  ORDERING_DIRECTION = 'orderingDirection',
```
Comment (pavel-romanov8) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:36]: Not sure why we are using snake case here

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 8]
```diff
@@ -30,12 +73,27 @@ export class AnimeTableComponent implements OnInit {
     'status',
   ] as const;
 
+  /** Anime type options map. */
+  public readonly animeTypeOptionsMap = animeTypeOptionsMap;
+
+  @ViewChild(MatSort) private tableSort?: MatSort;
```
Comment (pavel-romanov8) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:79]: We usually place a field under decorator like so:
```suggestion
  @ViewChild(MatSort)
  private tableSort?: MatSort;
```

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 9]
```diff
@@ -50,12 +108,186 @@ export class AnimeTableComponent implements OnInit {
 
   /** @inheritDoc */
   public ngOnInit(): void {
-    this.animeList$.pipe(
+    this.filtersForm.valueChanges.pipe(
+      untilDestroyed(this),
+    ).subscribe(
```
Comment (pavel-romanov8) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:113]: This configuration of `subscribe` already marked as deprecated in latest versions of RxJS. Instead, please, use it like so:
```ts
subscribe({
  next: () => {}
})
```

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 10]
```diff
@@ -50,12 +108,186 @@ export class AnimeTableComponent implements OnInit {
...
+      filterValues => {
+        const filterOptions: FilterOption<AnimeFilterOptions>[] = [];
+
+        const search = filterValues[AnimeFilterOptions.Search];
```
Comment (pavel-romanov8) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:117]: We are usually prefer to put side effects into `tap` operator. That way we are making them more explicit

File: apps/angular/src/core/rxjs/withIsFirst.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+/**
+ * Custom operator that provides isFirst parameter.
+ * `isFirst` says is this first emit.
+ * @param predicate Predicate in which isFirst parameter will be provided.
+ */
+export function withIsFirst<TItem, TReturn>(
+  predicate: (item: TItem, isFirst: boolean) => TReturn,
+): (source$: Observable<TItem>) => Observable<TReturn> {
+  let isFirst = true;
```
Comment (pavel-romanov8) [inline apps/angular/src/core/rxjs/withIsFirst.ts:11]: It looks like we are trying to introduce imperative style into declarative Rx. Instead of doing so, try to change the stream flow

File: libs/core/models/anime/animeType.ts [hunk 12]
```diff
@@ -35,6 +35,27 @@ export namespace AnimeType {
...
 }
+
+export const animeTypeOptionsMap: Readonly<Record<AnimeType, string>> = {
+  [AnimeType.TV]: AnimeType.toReadable(AnimeType.TV),
+  [AnimeType.OVA]: AnimeType.toReadable(AnimeType.OVA),
+  [AnimeType.Movie]: AnimeType.toReadable(AnimeType.Movie),
+  [AnimeType.Special]: AnimeType.toReadable(AnimeType.Special),
+  [AnimeType.ONA]: AnimeType.toReadable(AnimeType.ONA),
+  [AnimeType.Music]: AnimeType.toReadable(AnimeType.Music),
+  [AnimeType.Unknown]: AnimeType.toReadable(AnimeType.Unknown),
```
Comment (pavel-romanov8) [inline libs/core/models/anime/animeType.ts:60]: Please, create a pipe for that, instead of such dictionary

File: apps/angular/src/core/utils/table.ts [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+ * @param sort Material sort.
+ */
+export function matSortToSortOptions<TSortFields extends string>(sort: Sort): SortOptions<TSortFields> | null {
+  if (sort.direction === '') {
+    return null;
+  }
+
+  return {
+    field: <TSortFields>sort.active,
+    direction: <SortDirection>sort.direction,
```
Comment (pavel-romanov8) [inline apps/angular/src/core/utils/table.ts:15]: You are not converting, but casting it

File: apps/angular/src/core/utils/list-manager.ts [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+  >;
+
+  /** Is list loading. */
+  public readonly isLoading$: Observable<boolean>;
+
+  private readonly loading$ = new BehaviorSubject(true);
+
+  private readonly reload$ = new BehaviorSubject<void>(undefined);
+
+  private readonly pagination$ = new BehaviorSubject(new PaginationData());
```
Comment (pavel-romanov8) [inline apps/angular/src/core/utils/list-manager.ts:60]: If you are proving generics for other subjects let's provide one for this one as well

File: apps/angular/src/core/utils/list-manager.ts [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+export interface ListManagerInitParams<TItemSortFields, TItemFilterFields> {
+
+  /** Initial pagination data. */
+  paginationData?: PaginationData;
+
+  /** Initial sort options. */
+  sortOptions?: SortOptions<TItemSortFields>;
+
+  /** Initial filters options. */
+  filtersOptions?: readonly FilterOption<TItemFilterFields>[];
```
Comment (pavel-romanov8) [inline apps/angular/src/core/utils/list-manager.ts:36]: Are they mutable?

**Review by pavel-romanov8 — CHANGES_REQUESTED** [review]

**Review by dbitkovski — APPROVED** [review]

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 16]
```diff
@@ -50,12 +105,184 @@ export class AnimeTableComponent implements OnInit {
...
+
+    this.listManager.pagePagination$.pipe(
       untilDestroyed(this),
-    ).subscribe(
-      () => {
-        this.isLoading = false;
-      },
+      withLatestFrom(this.listManager.paginationResetParams$),
+      tap(([pagination, [sortOptions, filters]]) => {
+        const queryParams: Record<string, string> = {
```
Comment (pavel-romanov8) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:138]: I believe that we can use `TableQueryParams` as a key for the record instead of plain string

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 17]
```diff
@@ -50,12 +105,184 @@ export class AnimeTableComponent implements OnInit {
...
+          [TableQueryParams.Page]: pagination.page.toString(),
+          [TableQueryParams.PageSize]: pagination.pageSize.toString(),
+        };
+
+        queryParams[TableQueryParams.OrderingField] = sortOptions !== null ? sortOptions.field : '';
+        queryParams[TableQueryParams.OrderingDirection] = sortOptions !== null ? sortOptions.direction : '';
```
Comment (pavel-romanov8) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:144]: Not sure why we can initialize it directly in `queryParams` variable

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 18]
```diff
@@ -50,12 +105,184 @@ export class AnimeTableComponent implements OnInit {
...
+    }
+  }
+
+  /**
+   * Paginator changed.
+   * @param page Page event.
+   */
+  public onPaginationChanged(page: PageEvent): void {
+    this.listManager.updatePagination(new PaginationData(
+      page.pageIndex + 1,
```
Comment (pavel-romanov8) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:206]: By adding `+1` we are showing that the function knows something about the `PaginationData` inner implementation and because of that adding this +1 in order to make it works properly. Let's move this responsibility from the component to a class itself

File: libs/core/pagination.ts [hunk 19]
```diff
@@ -0,0 +1,24 @@
+/** Pagination data. */
+export class PaginationData {
```
Comment (pavel-romanov8) [inline libs/core/pagination.ts:2]: Not sure that `Data` brings much value here. We can omit it and name the class like _Pagination_

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 20]
```diff
@@ -50,12 +105,184 @@ export class AnimeTableComponent implements OnInit {
...
+      this.tableSort.sort(
+        {
+          id: sortOptions.field,
+          start: sortOptions.direction,
+          disableClear: false,
+        },
+      );
+
+      // Tricky way to activate styles for sorted column
+      (this.tableSort.sortables.get(sortOptions.field) as MatSortHeader)._setAnimationTransitionState(
```
Comment (pavel-romanov8) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:183]: Please, avoid usage of methods that starts with lower-dash because they are meant for internal usage. It is a common agreement in JS community

File: apps/angular/src/core/utils/queryParameters.ts [hunk 21]
```diff
@@ -0,0 +1,19 @@
+import { ParamMap } from '@angular/router';
+
+/**
+ * Get number from query parameters.
+ * @param paramsMap Parameters map.
+ * @param key Key of parameter.
+ */
+export function getNumberQueryParameter(paramsMap: ParamMap, key: string): number | undefined {
```
Comment (pavel-romanov8) [inline apps/angular/src/core/utils/queryParameters.ts:8]: Don't you think that we are having too many functionalities as _util_?

File: apps/angular/src/core/utils/list-manager.ts [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+  switchMap,
+  switchMapTo,
+  tap,
+  withLatestFrom,
+} from 'rxjs';
+import { PaginationData } from '@js-camp/core/pagination';
+import { SortOptions } from '@js-camp/core/models/sortOptions';
+import { FilterOption } from '@js-camp/core/models/filterOption';
+
+export type ApiRequestCallback<TItem, TItemSortFields, TItemFilterFields> = (
```
Comment (pavel-romanov8) [inline apps/angular/src/core/utils/list-manager.ts:19]: Looks like `ListManageer` is more about UI layer. If that is the case, we shouldn't use any terms from repository layer because it is a leak of abstraction. You can check the architecture of our apps in wiki in order to have a common understanding of what we are trying to achieve - https://wiki.saritasa.rocks/frontend/frameworks/architecture/

**Review by pavel-romanov8 — CHANGES_REQUESTED** [review]

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 23]
```diff
@@ -50,12 +105,184 @@ export class AnimeTableComponent implements OnInit {
...
+    if (filtersOptions !== undefined) {
+      for (const filterOption of filtersOptions) {
+        if (typeof filterOption.value === 'string' && filterOption.field === AnimeFilterOptions.Search) {
+          this.filtersForm.controls[filterOption.field].setValue(filterOption.value);
+        }
+        if (filterOption.field === AnimeFilterOptions.Type) {
+          this.filtersForm.controls[filterOption.field].setValue(filterOption.value as AnimeType[]);
+        }
+      }
+    }
```
Comment (chernodub) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:197]: Please consider reverting the conditions to remove the unnecessary indents and make the code more readable

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 24]
```diff
@@ -50,12 +105,184 @@ export class AnimeTableComponent implements OnInit {
...
+
+    if (filtersOptions !== undefined) {
+      for (const filterOption of filtersOptions) {
+        if (typeof filterOption.value === 'string' && filterOption.field === AnimeFilterOptions.Search) {
+          this.filtersForm.controls[filterOption.field].setValue(filterOption.value);
+        }
+        if (filterOption.field === AnimeFilterOptions.Type) {
+          this.filtersForm.controls[filterOption.field].setValue(filterOption.value as AnimeType[]);
+        }
```
Comment (chernodub) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:195]: This switchcase-like logic makes me think if we even need the generalization of filters, then? Can we just have a single object representing the whole filter? something like:

```ts
interface PaginationQuery {
  searchString: string;
  type: AnimeType[];
  page: number;
  pageSize: number;
  // ...
}
```

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 25]
```diff
@@ -50,12 +105,184 @@ export class AnimeTableComponent implements OnInit {
...
+        field: AnimeFilterOptions.Search,
+        filterType: FilterType.Exact,
+        value: searchParam,
+      });
+    }
+
+    return {
+      paginationData,
+      sortOptions,
+      filtersOptions,
```
Comment (chernodub) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:285]: After all, for me, it seems like you tried to generalize the filter params too much. Having an array of different filters won't bring much benefit, but instead (1) makes the code too hard to understand and (2) introduce some switchcase-like logic. I'd recommend simplifying it with some simple interface representing a query, like I suggested before.

File: apps/angular/src/core/services/anime.service.ts [hunk 26]
```diff
@@ -27,15 +35,39 @@ export class AnimeService {
...
+   * @param filterOptions Filter options.
+   */
+  public getAnimeList(
+    paginationData: PaginationData,
+    sortOptions: SortOptions<AnimeSortField> | null,
+    filterOptions: readonly FilterOption<AnimeFilterOptions>[] | null,
+  ): Observable<LimitOffsetPagination<AnimeBase>> {
+    let filterParams = {};
+    if (filterOptions !== null) {
+      filterParams = FilterOptionMap.toDto(filterOptions);
```
Comment (chernodub) [inline apps/angular/src/core/services/anime.service.ts:51]: I'd recommend to make the name more explicit `FilterOptionMap` -> `FilterOptionMapper`

File: apps/angular/src/core/utils/list-manager.ts [hunk 27]
```diff
@@ -0,0 +1,164 @@
...
+
+  /** Pagination options.*/
+  public readonly pagePagination$: Observable<PaginationData>;
+
+  /** Parameters that can reset pagination. */
+  public readonly paginationResetParams$: Observable<
+    [
+      SortOptions<TItemSortFields> | null,
+      readonly FilterOption<TItemFilterOptions>[] | null,
+    ]
```
Comment (chernodub) [inline apps/angular/src/core/utils/list-manager.ts:49]: We usually try not to use tuples in such context. This promotes other devs to decrease the readability of the code, since they basically have to give the names for the properties themselves, consider the following:
```ts
this.paginationResetParams$.pipe(
	map(([s, f]) => {...}),
);
```

Also, refactoring such thing would be a mess. E.g. for some reason, you'll need to change the order of these items, you'll have to refactor every place where it's used (rename the parameters).

tldr: please prefer objects over tuples to pass data. This approach makes your code more expressive

File: apps/angular/src/core/utils/list-manager.ts [hunk 28]
```diff
@@ -0,0 +1,164 @@
...
+  private readonly filters$ = new BehaviorSubject<readonly FilterOption<TItemFilterOptions>[] | null>(null);
+
+  public constructor({
+    paginationData,
+    sortOptions,
+    filtersOptions,
+  }: ListManagerInitParams<TItemSortFields, TItemFilterOptions> = {}) {
+    if (sortOptions !== undefined) {
+      this.sort$.next(sortOptions);
+    }
```
Comment (chernodub) [inline apps/angular/src/core/utils/list-manager.ts:72]: Instead of triggering subject multiple times (with initial value and then second initial in the constructor), prefer intializing it inline with the argument passed in constructor. This would reduce the cognitive complexity in your code

Comment (chernodub) [reply apps/angular/src/core/utils/list-manager.ts:72]: e.g:
```ts
this.sort$ = new BehaviorSubject(sortOptions ?? null)
``

File: apps/angular/src/core/utils/list-manager.ts [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+      shareReplay({ bufferSize: 1, refCount: true }),
+    );
+
+    this.pagePagination$ = merge(
+      this.pagination$,
+      this.paginationResetParams$.pipe(
+        skip(1),
+        withLatestFrom(this.pagination$),
+        map(([, pagination]) => this.resetPagination(pagination)),
+      ),
```
Comment (chernodub) [inline apps/angular/src/core/utils/list-manager.ts:98]: Please prefer moving it into constants instead of initializing such stuff inline. This would increase the readability of your code. Also, you can move the intiailization into functions, to make constructor more expressive

File: apps/angular/src/core/utils/list-manager.ts [hunk 30]
```diff
@@ -0,0 +1,164 @@
...
+      this.filters$,
+    ]).pipe(
+      debounceTime(500),
+      shareReplay({ bufferSize: 1, refCount: true }),
+    );
+
+    this.pagePagination$ = merge(
+      this.pagination$,
+      this.paginationResetParams$.pipe(
+        skip(1),
```
Comment (chernodub) [inline apps/angular/src/core/utils/list-manager.ts:95]: Usually, skipping value looks workaround-ish. For this case, I don't get why you're doing it. Please prefer either omitting such stuff from your code, or leaving a comment explaining why the first value is skipped. Because, for me, if observable called `resetParams$` it should reset params **every time** it emits.

File: apps/angular/src/core/utils/list-manager.ts [hunk 31]
```diff
@@ -0,0 +1,164 @@
...
+
+  /** Initial filters options. */
+  readonly filtersOptions?: readonly FilterOption<TItemFilterOptions>[];
+}
+
+/** Manager to work with lists. */
+export class ListManager<TItem, TItemSortFields, TItemFilterOptions> {
+
+  /** Pagination options.*/
+  public readonly pagePagination$: Observable<PaginationData>;
```
Comment (chernodub) [inline apps/angular/src/core/utils/list-manager.ts:42]: Seems tautological/non-representative, can we improve the name?

File: apps/angular/src/core/utils/list-manager.ts [hunk 32]
```diff
@@ -0,0 +1,164 @@
...
+  ): Observable<readonly TItem[]> {
+    return this.reload$.pipe(
+      switchMapTo(this.paginationResetParams$),
+      withLatestFrom(this.pagePagination$),
+      tap(() => this.loading$.next(true)),
+      switchMap(([[sortOptions, filterOptions], pagination]) => fetchFunc(pagination, sortOptions, filterOptions)),
+      tap(() => this.loading$.next(false)),
+      tap(paginatedList => {
+        this.updatePagination(paginatedList.pagination);
+      }),
```
Comment (chernodub) [inline apps/angular/src/core/utils/list-manager.ts:119]: What's the point updating it second time if it's already changed in the first call of `updatePagination`? Seems like it's almost a hacky and implicit loop

File: apps/angular/src/core/utils/list-manager.ts [hunk 33]
```diff
@@ -0,0 +1,164 @@
...
+
+  /**
+   * Get paginated items.
+   * @param fetchFunc Function to fetch items.
+   */
+  public getPaginatedItems(
+    fetchFunc: ApiRequestCallback<TItem, TItemSortFields, TItemFilterOptions>,
+  ): Observable<readonly TItem[]> {
+    return this.reload$.pipe(
+      switchMapTo(this.paginationResetParams$),
```
Comment (chernodub) [inline apps/angular/src/core/utils/list-manager.ts:112]: Since you wait for both to be triggered, why not generalizing the logic into a single subject?

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 34]
```diff
@@ -50,12 +105,184 @@ export class AnimeTableComponent implements OnInit {
...
+  /**
+   * Paginator changed.
+   * @param page Page event.
+   */
+  public onPaginationChanged(page: PageEvent): void {
+    this.listManager.updatePagination(new PaginationData(
+      page.pageIndex + 1,
+      page.pageSize,
+      page.length,
+    ), true);
```
Comment (chernodub) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:209]: This call seems to be unreadable because of the second parameter. https://wiki.saritasa.rocks/frontend/languages/ts-js/best-practices#how-to-pass-arguments-to-functionsmethods

File: libs/core/pagination.ts [hunk 35]
```diff
@@ -0,0 +1,24 @@
...
+  /** Current page. */
+  public page: number;
+
+  /** Page size. */
+  public pageSize: number;
+
+  /** Total count. */
+  public totalCount: number;
+
+  private defaultPageSize = 10;
```
Comment (chernodub) [inline libs/core/pagination.ts:12]: 1. Since it's never reassigned, prefer `readonly`
2. Since it's a constant, prefer making it module scoped and https://wiki.saritasa.rocks/frontend/languages/ts-js/naming#naming-convention

In general, please always try to reduce the complexity of your classes/interfaces

File: apps/angular/src/core/utils/pagination.ts [hunk 36]
```diff
@@ -0,0 +1,13 @@
...
+/**
+ * Convert pagination data to limit offset options.
+ * @param paginationData Pagination data.
+ */
+export function paginationDataToLimitOffsetOptions(paginationData: PaginationData): LimitOffsetPaginationOptions {
+  return {
+    limit: paginationData.pageSize,
+    offset: paginationData.offset,
+  };
+}
```
Comment (chernodub) [inline apps/angular/src/core/utils/pagination.ts:13]: I don't see it used anywhere in this PR. Also, this seems like a mapper from domain to DTO, probably it's not a good idea having this in utils folder

File: apps/angular/src/core/utils/list-manager.ts [hunk 37]
```diff
@@ -0,0 +1,164 @@
...
+} from 'rxjs';
+import { PaginationData } from '@js-camp/core/pagination';
+import { SortOptions } from '@js-camp/core/models/sortOptions';
+import { FilterOption } from '@js-camp/core/models/filterOption';
+
+export type ApiRequestCallback<TItem, TItemSortFields, TItemFilterFields> = (
+  paginationData: PaginationData,
+  sortOptions: SortOptions<TItemSortFields> | null,
+  filterOptions: readonly FilterOption<TItemFilterFields>[] | null,
+) => Observable<LimitOffsetPagination<TItem>>;
```
Comment (chernodub) [inline apps/angular/src/core/utils/list-manager.ts:23]: Like @pavel-romanov8  mentioned, it's not only about the name. Your BE-provided representation of pagination leaks too. When operating on domain layer, we basically want to abstract ourselves from `limit`/`offset` but only use `pageNumber` and `pageSize`

File: apps/angular/src/core/utils/queryParameters.ts [hunk 38]
```diff
@@ -0,0 +1,19 @@
...
+  const stringParam = paramsMap.get(key);
+  if (stringParam === null) {
+    return undefined;
+  }
+  const numberParam = parseInt(stringParam, 10);
+
+  if (Number.isNaN(numberParam)) {
+    return undefined;
+  }
+  return numberParam;
```
Comment (chernodub) [inline apps/angular/src/core/utils/queryParameters.ts:18]: This could be reduced to

```suggestion
const value = parseInt(paramsMap.get(key), 10);
return isNaN(value) ? undefined : value;
```

File: libs/core/models/sortOptions.ts [hunk 39]
```diff
@@ -1,7 +1,15 @@
...
+  Ascending = 'asc',
+  Descending = 'desc',
+}
+
+/**
+ * Check if string is sort direction.
+ * @param sortDirection String to check.
+ */
+export function isSortDirection(sortDirection: string): sortDirection is SortDirection {
+  return Object.values(SortDirection).includes(sortDirection as SortDirection);
```
Comment (chernodub) [inline libs/core/models/sortOptions.ts:12]: Same here with converting enum to array using `Object.values`, be careful and consider using library I recommended before

File: libs/core/pagination.ts [hunk 40]
```diff
@@ -0,0 +1,24 @@
...
+  public constructor(page?: number, pageSize?: number, totalCount?: number) {
+    this.page = page ?? 1;
+    this.pageSize = pageSize ?? this.defaultPageSize;
+    this.totalCount = totalCount ?? 1;
+  }
+
+  /** Get offset. */
+  public get offset(): number {
+    return this.pageSize * (this.page - 1);
+  }
```
Comment (chernodub) [inline libs/core/pagination.ts:23]: Do we need this info on domain layer? I believe no, this should be isolated in the mapping to DTO representation

File: libs/core/pagination.ts [hunk 41]
```diff
@@ -0,0 +1,24 @@
...
+  /** Page size. */
+  public pageSize: number;
+
+  /** Total count. */
+  public totalCount: number;
+
+  private defaultPageSize = 10;
+
+  public constructor(page?: number, pageSize?: number, totalCount?: number) {
+    this.page = page ?? 1;
```
Comment (chernodub) [inline libs/core/pagination.ts:15]: Is your pagination 1-based? If so, consider mentioning this in the docs. I believe, usually devs expect 0-based, if they're not R-fans

**Review by chernodub — CHANGES_REQUESTED** [review]

File: apps/angular/src/app/features/anime/table/anime-table.component.ts (hunk 20 above)
Comment (aletor123) [reply apps/angular/src/app/features/anime/table/anime-table.component.ts:183]: This need because of issue
https://github.com/angular/components/issues/10242

File: apps/angular/src/core/utils/queryParameters.ts (hunk 21 above)
Comment (aletor123) [reply apps/angular/src/core/utils/queryParameters.ts:8]: I thinks it's okay to store reusable functions in utils

File: apps/angular/src/core/utils/list-manager.ts (hunk 32 above)
Comment (aletor123) [reply apps/angular/src/core/utils/list-manager.ts:119]: This need to set total count from response

File: apps/angular/src/core/utils/list-manager.ts (hunk 33 above)
Comment (aletor123) [reply apps/angular/src/core/utils/list-manager.ts:112]: Because this will trigger infinite reload because of updating pagination after response

Comment (aletor123) [reply apps/angular/src/core/utils/list-manager.ts:119]: This need because i'm updating pagination after response and this mustn't trigger in this situation

File: apps/angular/src/app/features/anime/table/anime-table.component.ts (hunk 20 above)
Comment (chernodub) [reply apps/angular/src/app/features/anime/table/anime-table.component.ts:158]: Probably you could initialize the value on template then? https://material.angular.io/components/sort/api#:~:text=The%20direction%20to%20set%20when%20an%20MatSortable%20is%20initially%20sorted.%20May%20be%20overridden%20by%20the%20MatSortable%27s%20sort%20start

File: apps/angular/src/core/services/anime.service.ts [hunk 42]
```diff
@@ -27,15 +27,50 @@ export class AnimeService {
...
+          value: filterOptions.searchString,
+        },
+      ]);
+    }
     return this.http.get<LimitOffsetPaginationDto<AnimeDto>>(
       this.appUrls.animeUrls.list,
+      {
+        params: {
+          ...LimitOffsetQueryMapper.toDto(
+            LimitOffsetPaginationMapper.mapPaginationToLimitOffsetOptions(pagination),
```
Comment (chernodub) [inline apps/angular/src/core/services/anime.service.ts:61]: Why not call it inside `LimitOffsetQueryMapper` to fully abstract the mapping to dto?

File: apps/angular/src/app/features/anime/table/anime-table.component.ts [hunk 43]
```diff
@@ -50,12 +99,153 @@ export class AnimeTableComponent implements OnInit {
...
+          types: type != null ? type : [],
+        };
+
+        this.listManager.updateFilters(filterOptions);
+      }),
+      untilDestroyed(this),
+    ).subscribe();
+
+    this.listManager.pagination$.pipe(
+      withLatestFrom(this.listManager.paginationResetParams$),
```
Comment (chernodub) [inline apps/angular/src/app/features/anime/table/anime-table.component.ts:115]: `this.listManager.paginationResetParams$` seems like an observable that should be isolated inside `listManager`, because the name tells me that this observable emits when I need to reset pagination. But it seems that you're trying to sit on two chairs at the same time. I believe it would be a good idea to either simply expose `sortOptions$` and `filterOptions$` observables, or just rename the property

**Review by chernodub — APPROVED** [review]
